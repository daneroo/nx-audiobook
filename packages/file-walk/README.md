# file-walk

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test file-walk` to execute the unit tests via [Jest](https://jestjs.io).

## Running lint

Run `nx lint file-walk` to execute the lint via [ESLint](https://eslint.org/).
