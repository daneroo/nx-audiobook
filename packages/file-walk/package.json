{
    "name": "@nx-audiobook/file-walk",
    "version": "1.0.0",
    "description": "Filesystem traversal",
    "type": "module",
    "module": "./dist/main.js",
    "types": "./dist/main.d.ts",
    "exports": {
        ".": "./dist/main.js"
    },
    "files": [
        "dist",
        "!dist/.tsbuildinfo"
    ],
    "scripts": {
        "lint": "eslint .",
        "test": "vitest run",
        "coverage": "vitest run --coverage",
        "build": "pnpm tsc --build ."
    },
    "license": "ISC",
    "dependencies": {
        "@root/walk": "^1.1.0"
    }
}