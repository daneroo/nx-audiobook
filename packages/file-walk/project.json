{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/file-walk/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/file-walk/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/file-walk"],
      "options": {
        "jestConfig": "packages/file-walk/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
