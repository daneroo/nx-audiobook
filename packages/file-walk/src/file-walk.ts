import { walk, type WalkFunc } from '@root/walk'

import { basename, extname } from 'node:path'
import { promises as fs } from 'node:fs'

// Returns all subdirectories of rootPath, recursively, including rootPath itself
export async function getDirectories(rootPath: string): Promise<string[]> {
  const directories: string[] = []
  // eslint-disable-next-line @typescript-eslint/require-await
  const walker: WalkFunc = async (err, pathname, dirent) => {
    if (err !== null && err !== undefined) {
      // throw an error to stop walking (or return to ignore and keep going)
      console.warn('fs stat error for %s: %s', pathname, err.message)
      return false
    }
    if (dirent.isDirectory()) {
      directories.push(pathname)
      return true
    }
    return false
  }

  await walk(rootPath, walker)
  return directories
}

export interface FileInfo {
  path: string
  basename: string
  extension: string
  size: number
  mtime: Date
}

// Returns all files (as File Objects) from rootPath
// if options.recurse is false, only get files in rootPath, not subdirectories
// if options.stat is false, only get path extension and name, not size or mtime
export async function getFiles(
  rootPath: string,
  options: { recurse: boolean; stat: boolean } = { recurse: false, stat: false }
): Promise<FileInfo[]> {
  const files: FileInfo[] = []

  const walker: WalkFunc = async (err, pathname, dirent) => {
    if (err !== null && err !== undefined) {
      // throw an error to stop walking (or return to ignore and keep going)
      console.warn('fs stat error for %s: %s', pathname, err.message)
      return false
    }
    if (dirent.isDirectory()) {
      // We recurse into subdirectories when either
      // - recurse parameter is true
      // - we are at the root
      return options.recurse || rootPath === pathname
    } else if (dirent.isFile()) {
      const { size, mtime } = options.stat
        ? await fs.stat(pathname)
        : { size: 0, mtime: new Date(0) }
      const file: FileInfo = {
        path: pathname,
        basename: basename(pathname),
        extension: extname(pathname),
        size,
        mtime,
      }
      files.push(file)
    }
    return false
  }
  await walk(rootPath, walker)
  return files
}

// Return a single FileInfo object for the directory at pathname
export async function getDirectory(pathname: string): Promise<FileInfo> {
  const { size, mtime } = await fs.stat(pathname)
  const file: FileInfo = {
    path: pathname,
    basename: basename(pathname),
    extension: extname(pathname),
    size,
    mtime,
  }
  return file
}
