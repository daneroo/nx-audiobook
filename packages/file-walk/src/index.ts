export * from './lib/file-walk';
