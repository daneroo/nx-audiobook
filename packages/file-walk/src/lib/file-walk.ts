export function fileWalk(): string {
  return 'file-walk';
}
