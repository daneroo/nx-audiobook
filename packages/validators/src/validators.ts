import type { FileInfo } from '@nx-audiobook/file-walk'

import type { Validation } from './types'

// for files in a set (typically a directory),
// verify that all extensions (and some known filenames are accounted for)
// simply console.error the unaccounted for files files.
// three inputs options:
//  audio file extensions: e.g. ['.mp3', '.m4b', '.m4a']
//  allowed (but not yet required filenames): e.g. ['coder.jpg', 'cover.png', 'metadata.json']
//  ignored (known, non-audio) file extensions: e.g. ['.jpg', '.png', '.pdf', '.epub']
//  ignored (known, non-audio) filenames: e.g. ['.DS_Store', 'MD5SUM']

export const audioExtensions = ['.mp3', '.m4b', '.m4a']
export const ebookExtensions = ['.pdf', '.epub']
export const allowedFilenames = ['cover.jpg', 'cover.png', 'metadata.json']
export const ignoredExtensions = ['.tiff']
export const ignoredFilenames = ['.DS_Store', 'MD5SUM']

export function isAudioFile(fileInfo: FileInfo): boolean {
  return audioExtensions.includes(fileInfo.extension)
}

export function isEBookFile(fileInfo: FileInfo): boolean {
  return ebookExtensions.includes(fileInfo.extension)
}
export function isAllowed(fileInfo: FileInfo): boolean {
  return allowedFilenames.includes(fileInfo.basename)
}

export function isIgnored(fileInfo: FileInfo): boolean {
  return (
    ignoredFilenames.includes(fileInfo.basename) ||
    ignoredExtensions.includes(fileInfo.extension)
  )
}

export function validateFilesAllAccountedFor(files: FileInfo[]): Validation {
  // count ignored files
  const ignored = files.filter(isIgnored).length

  // count audio files
  const audio = files.filter(isAudioFile).length

  // count ebook files
  const ebook = files.filter(isEBookFile).length

  // count allowed named files
  const allowed = files.filter(isAllowed).length

  // the actual list of unaccounted for files (not audio , not ignored (ext or basename))
  const unaccounted = files
    .filter((fileInfo) => {
      // *not* (audio or  ignored (ext or basename))
      return !(
        isAudioFile(fileInfo) ||
        isEBookFile(fileInfo) ||
        isAllowed(fileInfo) ||
        isIgnored(fileInfo)
      )
    })
    // now extract the path
    .map((fileInfo) => fileInfo.path)
  const ok = unaccounted.length === 0

  const extra = {
    total: files.length,
    audio,
    ebook,
    allowed,
    ignored,
    unaccounted,
  }

  const validation: Validation = {
    ok,
    level: ok ? 'info' : 'warn',
    message: ok ? 'All accounted for' : 'Have unaccounted for files',
    extra,
  }
  return validation
}
