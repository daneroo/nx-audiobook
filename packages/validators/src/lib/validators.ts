import { Validation } from './types';

import {
  filterAudioFileExtensions,
  // filterNonAudioFileExtensions,
  filterNonAudioExtensionsOrNames,
  // filterNonAudioFilenames,
} from './filters';
import { File } from '@nx-audiobook/file-walk';

// for filenames in a set (typically a directory),
// verify that all extensions (and some known filenames are accounted for)
// simply console.error the unaccounted for files files.
export function validateFilesAllAccountedFor(files: File[]): Validation {
  const filenames = files.map((file) => file.path);
  const excludedFilenames = filenames.filter(filterNonAudioExtensionsOrNames);
  const audioFiles = filenames.filter(filterAudioFileExtensions);
  const unclassified = filenames.filter((filePath) => {
    if (filterAudioFileExtensions(filePath)) return false;
    if (filterNonAudioExtensionsOrNames(filePath)) return false;
    return true;
  });
  const ok = unclassified.length === 0;
  const extra = {
    total: filenames.length,
    excluded: excludedFilenames.length,
    audio: audioFiles.length,
    unclassified: unclassified,
  };

  const validation: Validation = {
    ok,
    level: ok ? 'info' : 'warn',
    message: ok ? 'All accounted for' : 'Have unclassified files',
    extra,
  };
  return validation;
}
