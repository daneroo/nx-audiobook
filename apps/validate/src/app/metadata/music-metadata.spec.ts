import { describe, expect, it, beforeEach } from 'vitest'
import * as fs from 'node:fs/promises'
import * as path from 'node:path'
import { getFiles } from '@nx-audiobook/file-walk'
import { isAudioFile } from '@nx-audiobook/validators'

import {
  AudioBookMetadata,
  getMetadataForSingleFile,
  fetchResult,
} from './music-metadata'

const assetBaseDirectory = path.resolve(
  __dirname,
  '../../../../..',
  'assets',
  'audio'
)

describe('audio assets', () => {
  it('should find assets directory', async () => {
    expect(
      (await fs.readdir(assetBaseDirectory)).filter((f) => f !== '.DS_Store')
    ).toMatchInlineSnapshot(`
      [
        "Derek Walcott - Love After Love",
        "Robert Frost - The Road not Taken",
      ]
    `)
  })
  it('should find audio files', async () => {
    expect(
      (await getFiles(assetBaseDirectory, { recurse: true, stat: false }))
        .map((fi) => fi.basename)
        .filter((basename) => basename !== '.DS_Store')
    ).toMatchInlineSnapshot(`
      [
        "Derek Walcott - Love After Love.mp3",
        "Robert Frost - The Road not Taken.mp3",
      ]
    `)
  })
})

// This adds the proper type for process.env.CACHE_BASE_DIRECTORY, so we can use the dot notation and delete operator
declare let process: { env: { CACHE_BASE_DIRECTORY?: string } }

const cacheBaseDirectory = path.join(__dirname, 'test_cache')
beforeEach(async () => {
  try {
    await fs.rm(cacheBaseDirectory, { recursive: true })
    process.env.CACHE_BASE_DIRECTORY = cacheBaseDirectory
  } catch (error) {}

  // clean up function, called once after each test run
  return async () => {
    try {
      await fs.rm(cacheBaseDirectory, { recursive: true })
      delete process.env.CACHE_BASE_DIRECTORY
    } catch (error) {}
  }
})

describe('happy path', () => {
  it('should find audio file metadata', async () => {
    const fileInfos = (
      await getFiles(assetBaseDirectory, { recurse: true, stat: true })
    ).filter(isAudioFile)
    expect(fileInfos).toHaveLength(2)

    const metas: AudioBookMetadata[] = []
    for (const fileInfo of fileInfos.filter(isAudioFile)) {
      metas.push(await getMetadataForSingleFile(fileInfo))
    }
    expect(metas).toMatchSnapshot()
  })
  it('should find audio file metadata (no cache)', async () => {
    const fileInfos = (
      await getFiles(assetBaseDirectory, { recurse: true, stat: true })
    ).filter(isAudioFile)
    expect(fileInfos).toHaveLength(2)

    const metas: AudioBookMetadata[] = []
    for (const fileInfo of fileInfos.filter(isAudioFile)) {
      metas.push(
        await fetchResult({
          fileInfo,
          options: {
            duration: false,
            includeChapters: false,
          },
        })
      )
    }
    expect(metas).toMatchSnapshot()
  })
})
