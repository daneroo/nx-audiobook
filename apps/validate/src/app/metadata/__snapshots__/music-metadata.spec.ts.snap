// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`happy path > should find audio file metadata (no cache) 1`] = `
[
  {
    "author": "Jon Kabat-Zinn, American Public Media",
    "cover": {
      "format": "image/png",
      "size": 650759,
    },
    "duration": 44.85224489795918,
    "title": "Speaking of Faith from American Public Media",
    "warning": {},
  },
  {
    "author": "Robert Frost",
    "cover": {
      "format": "image/jpeg",
      "size": 93035,
    },
    "duration": 76.19918367346939,
    "title": "Poem of the Day from Poetry Foundation.org",
    "warning": {},
  },
]
`;

exports[`happy path > should find audio file metadata 1`] = `
[
  {
    "author": "Jon Kabat-Zinn, American Public Media",
    "cover": {
      "format": "image/png",
      "size": 650759,
    },
    "duration": 44.85224489795918,
    "title": "Speaking of Faith from American Public Media",
    "warning": {},
  },
  {
    "author": "Robert Frost",
    "cover": {
      "format": "image/jpeg",
      "size": 93035,
    },
    "duration": 76.19918367346939,
    "title": "Poem of the Day from Poetry Foundation.org",
    "warning": {},
  },
]
`;
